File: interface/fortran/solution/evolve.f90
subroutine evolve(u, u_previous, nx, ny, a, dt, dx2, dy2)

   implicit none

   real*8, intent(inout) :: u(nx,ny), u_previous(nx, ny)
   !f2py intent(in, out) :: u, u_previous
   integer :: nx, ny
   real*8 :: a, dt, dx2, dy2

   integer :: i,j

   do j = 2, ny-1
     do i = 2, nx-1
        u(i,j) = u_previous(i,j) + a * dt * (  &
                 (u_previous(i-1, j) - 2*u_previous(i,j) +  &
                     u_previous(i+1, j)) / dx2 + &
                 (u_previous(i, j-1) - 2*u_previous(i,j) +  &
                     u_previous(i, j+1)) / dy2 )
     end do
   end do
   u_previous(:,:) = u(:,:)


end subroutine

File: interface/c/evolve.c
#include <string.h>

void evolve(double *u, double *u_previous, int nx, int ny, 
            double a, double dt, double dx2, double dy2)
{
  int i, j;
  int ij, ip, im, jp, jm;

  for (i=1; i < nx-1; i++)
    for (j=1; j < ny-1; j++) {
      // Linearisation for 2D array
      ij = j + i*nx;
      ip = j + (i+1)*nx;
      im = j + (i-1)*nx;
      jp = (j + 1) + i*nx;
      jm = (j - 1) + i*nx;
      u[ij] = u_previous[ij] + a * dt * ( 
              (u_previous[ip] - 2*u_previous[ij] + u_previous[im]) / dx2 + 
              (u_previous[jp] - 2*u_previous[ij] + u_previous[jm]) / dy2 );
      }

  memcpy(u_previous, u, nx*ny*sizeof(double));
}
                

